DEP: Deprecate scipy.misc in favour of scipy.datasets

AnirudhDagar opened the pull request:
#### Reference issue
Closes #15608

#### What does this implement/fix?
Completely deprecate `scipy.misc`.

TODO:
- [x] Replace `scipy.stats.rv_continuous` and `scipy.stats.kolmogn_p` usage of `misc.derivative`

#### Question
As discussed in the mailing list `central_diff_weights` and `derivative` should be deprecated and completely removed in the future. Within SciPy, `misc.derivative` is used internally only in `scipy.stats.rv_continuous` [here](https://github.com/scipy/scipy/blob/e3cd846ef353b10cc66972a5c7718e80948362ac/scipy/stats/_distn_infrastructure.py#L2069) and in `scipy.stats._ksstats._kolmogn_p` [here](https://github.com/scipy/scipy/blob/8dec95c7597fbeba43e212e40c7ef455db02088b/scipy/stats/_ksstats.py#L472).

~I'm unsure about what is the best way to replace that, since `misc.derivative` will be deprecated. Any suggestions will be helpful.~ Thanks in advance!

#### Additional information
ref #15607

h-vetinari commented:
Without preference one way or another, just noting that the more we explicitly announce removal of something by scipy 2.0.0, the more we create an expectation for this to arrive reasonably soon (i.e. it would be weird as a user IMO to keep seeing this warning for 5+ years).

Seeing the discussion in https://github.com/scipy/scipy/issues/15528, the impression I got is that many people see 2.0 still quite far on the horizon (though not never) compare also the notes in the [wiki](https://github.com/scipy/scipy/wiki/Ideas-for-changes-&-cleanups-in-SciPy-2.0) on this.

In the context of https://github.com/scipy/scipy/issues/14360, some modules now already announce a removal by 2.0.0, some don't.

rgommers reviewed (commented):
Thanks @AnirudhDagar! This looks pretty straightforward, but looks like it should wait till the `scipy.datasets` PR goes in first.

> I'm unsure about what is the best way to replace that, since `misc.derivative` will be deprecated.

I'd say it's best to just keep using it and suppress the deprecation warning. That decouples this from any functional changes to `scipy.stats`. When the time comes to remove `scipy.misc`, the person doing the cleanup can just move `derivative` to a private function inside `stats`, if the usage hasn't been cleaned up before then.

> Without preference one way or another, just noting that the more we explicitly announce removal of something by scipy 2.0.0, the more we create an expectation for this to arrive reasonably soon

There's no timeline at the moment and little pressure, the "we remove a module that may break a lot of older code" is just for a major version bump, which is all this warning says.

j-bowhay commented:
`central_diff_weights` and `derivative` state they will be removed in 2.0.0. `ascent`, `face` and `electrocardiogram` don't have a timeline for execution. I appreciate the whole submodule will be removed at 2.0.0 but without a timeline it is unclear whether these 3 will be left until 2.0.0 or removed after the usual two releases.

AnirudhDagar commented:
I guess I should also bump up the deprecation warnings to `1.10` since `1.9` is already underway unless we somehow manage to sneak both these PRs in before `1.9` release which I don't see happening.

AnirudhDagar commented:
```
FAILED ..\..\stats\tests\test_qmc.py::TestPoisson::test_mindist - numpy.core....
```

Test failure seems unrelated, but I'm curious if it's just a flake or an actual recurring problem in `main`.

rgommers commented:
@ngam yes indeed. This was discussed in [this mailing list thread](https://mail.python.org/archives/list/scipy-dev@python.org/thread/AACMUJES2D3F24QWSQNOBVIYBOUZD2QX/#AACMUJES2D3F24QWSQNOBVIYBOUZD2QX). There are other packages which are much better than these two fairly random functions, so best to point folks to those.

> Or we can obviously add a "differentiate" subpackage if there is sufficient use. Differentiation, like integration, are likely pretty important for users and so expanding the usage of derivative can be nice.

This has been discussed on and off for many years now, but no one has actually done the work (a few half-attempts failed). For example, see gh-2035 and gh-7457. At this point I think the topic is dead. If someone writes a high-quality module outside of SciPy and then wants to incorporate it, that can be discussed. But writing from scratch directly in the SciPy repo is no longer an option.

rgommers commented:
This is unblocked now that the `scipy.datasets` PR is merged.

AnirudhDagar commented:
Yes, I believe there is nothing to change in this PR though. Should be ready to go.

rgommers commented:
It looks like this deprecation warning shows up and should be filtered out so the doc build doesn't fail in CI:
```
DeprecationWarning:     `face` is deprecated!
    scipy.misc.face has been deprecated in SciPy v1.10.0; and will be completely removed in SciPy v1.12.0. Dataset methods have moved into the scipy.datasets module. Use scipy.datasets.face instead.
```

AnirudhDagar commented:
Sorry if this is a trivial question, but I don't know how to ignore the warning from the docstring example. I mean using warnings context manager in the docstring example is definitely not the correct way. I'm sure there is some other way to do that. @rgommers could you please point me to that? Thanks

rgommers commented:
You can add it to `filterwarnings` in `pytest.ini`, I'm fairly sure that's the correct way to suppress warnings like these.

AnirudhDagar commented:
Ok, so I noticed that I forgot about updating multiple other docstring examples that were using `scipy.misc` still. I've replaced those with `scipy.datasets` now.

Apartment from that, should I just filter out these two cases with a warnings content manager?
1. https://github.com/scipy/scipy/blob/59dac8a9fa9ea856f4a50521d295a3497d648faa/scipy/stats/_ksstats.py#L472
2. https://github.com/scipy/scipy/blob/59dac8a9fa9ea856f4a50521d295a3497d648faa/scipy/stats/_distn_infrastructure.py#L2018

ev-br commented:
Better not add an extra layer of pytest skips. Ideally, remove scipy.misc from the list of submodules to doctest, add scipy.datasets (the latter is a good move regardless of the former) https://github.com/scipy/scipy/blob/main/tools/refguide_check.py#L89.

Alternatively, add a filterwarings filter in refguide-check (gh-16391 has a better way of adding these kinds of filters)

ev-br commented:
> We can add a private derivative method 

I'd suggest to copy-paste it to scipy stats as a private function now and use it.

AnirudhDagar commented:
> I'd suggest to copy-paste it to scipy stats as a private function now and use it.

Thanks @ev-br where exactly in `scipy.stats` would you suggest moving the method to, so that it can be shared for both the above-mentioned cases?

> Ideally, remove scipy.misc from the list of submodules to doctest, add scipy.datasets (the latter is a good move regardless of the former)

Datasets submodule is already in the list, it was added in #15607. Removing misc from the list should do the trick.

AnirudhDagar commented:
Fixed the failing problems. This is ready for review, and I hope CI is also green now.

Ps. Added global deprecation doc warning for the documentation on `scipy.misc` page.

AnirudhDagar commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -6445,8 +6445,10 @@ def test_only__pdf(self):
         # _pdf is determined from _cdf by taking numerical derivative
-        dummy_distr = _distr2_gen(name='dummy')
-        assert_almost_equal(dummy_distr.pdf(1, a=1), 1)
+        with suppress_warnings() as sup:
+            sup.filter(DeprecationWarning)
```
Sorry missed this, can be removed now that we use a private method for `derivative`.

rgommers commented on scipy/stats/_common.py:
```diff
@@ -1,6 +1,140 @@
 ConfidenceInterval. __doc__ = "Class for confidence intervals."
+
+
+def _central_diff_weights(Np, ndiv=1):
```
It looks like it would be better to move this to `_lib` (e.g., `scipy/_lib/_finite_differences.py`), and then use it from both `stats` and `misc` - that way the code itself isn't duplicated. Also, `_lib` is a good place for this code to live, since it's not specific to `stats`. We have finite differences code in `optimize` too that may need dealing with at some point, and that'd be easier if it's in `_lib`.

AnirudhDagar replied:
Made the move to `scipy/_lib/_finite_differences.py` and now using that everywhere else to avoid duplication.

rgommers commented on scipy/misc/_common.py:
```diff
@@ -3,12 +3,19 @@
 
+@_deprecated(msg="scipy.misc.central_diff_weights is deprecated in "
+                 "SciPy v1.10.0; and will be completely removed in "
+                 "SciPy v2.0.0. You may consider using "
```
This should say `v1.12.0` just like the functions that were moved to `datasets`.

rgommers replied:
And same comment for `derivative`.

AnirudhDagar replied:
Done

rgommers commented on scipy/stats/_common.py:
```diff
@@ -1,6 +1,140 @@
+        raise ValueError("The number of points must be odd.")
+    from scipy import linalg
+    ho = Np >> 1
+    x = arange(-ho,ho+1.0)
```
The linter is quite unhappy with all this code. Now that it's being moved to a new file, may as well run `black` over the file? Or at least `autopep8`?

AnirudhDagar replied:
Ran black with 79 char line limit. Should be fixed.

rgommers reviewed (changes requested):
Thanks @AnirudhDagar. Overall this looks good. The main comment I have is that `from scipy import misc` does not actually raise a deprecation warning. It should, otherwise it's too easy to miss and we are likely to break more code when we actually remove it.

To do this and keep things working now that there's two levels of deprecated public-like names (ex: `misc` and `misc.common`), you can copy how this works in for example `scipy/sparse/linalg/eigen.py` and the `__init__.py` next to it (that had the same issue with `sparse.linalg.eigen` and `sparse.linalg.eigen.arpack`).

AnirudhDagar commented:
@rgommers sorry for the delay, I've made the changes as requested.

I was unsure what you meant by the similarity with `scipy/sparse/linalg/eigen.py` example. Please let me know if you meant something else regarding the deprecation of the whole misc module. Thanks!

AnirudhDagar commented:
Gentle ping @rgommers. Request you to review this again :) Thanks!

rgommers reviewed (approved):
Thanks @AnirudhDagar, LGTM now. In it goes!

> Please let me know if you meant something else regarding the deprecation of the whole misc module.

I actually meant that `from scipy import misc` by itself should raise a warning. But let's leave it as is for now. It's fairly annoying/disruptive to implement, and what is an easier strategy is to remove all the functions by `1.12.0` and then raise the warning once there are no more tests to worry about.

rgommers commented:
@AnirudhDagar could you add an entry for this in https://github.com/scipy/scipy/wiki/Release-Note-Entries-for-SciPy-1.10.0?

j-bowhay commented:
@h-vetinari there are a few things here slated for removal in 1.12.0 could you add them to the tracker please

h-vetinari commented:
> @h-vetinari there are a few things here slated for removal in 1.12.0 could you add them to the tracker please

Hey @j-bowhay, sorry I've been MIA for quite a while. 😑

I now added `scipy.misc` to the tracker (and tried to go through my remaining github notification emails) - please let me know if I missed something.
